docs/proposals/build-hooks.md

```diff
@@ -0,0 +1,218 @@
+# Build Hooks
+
+## Abstract
+
+A proposal for supporting the execution of hooks in the build lifecycle.
+
+
+## Motivation
+
+Users need to execute custom actions at different stages of the build process.
+OpenShift provides full flexibility on how builds are performed through the
+*Custom* build strategy, and *Source-To-Image* builds also allow customization
+via its `assemble` script. However, as we'll see, those extensibility points
+have their limitations.
+
+At the time of writing, **Deployments** have a hook mechanism that supports the
+execution of arbitrary commands before and after a deployment occurs,
+individually supported by each *Deployment Strategy*, namely *Recreate* and
+*Rolling*.
+The third existing strategy, *Custom*, does not support hooks.
+The implementation is done in the strategy level as to allow different
+strategies to support a potentially different set of hooks.
+
+Goals of this design:
+
+1. Identify build hook use cases
+2. Discuss and describe implementation details
+
+
+## Use cases
+
+### 1. Build pipeline
+
+As an application developer, I want to be able to [trigger a build after an
+upstream build completes](https://github.com/openshift/origin/issues/1228#issue-59933179),
+creating a build pipeline.
+
+
+### 2. Integration with external system
+
+As an application developer, I want to be notified or notify other people after
+a build completes, email or otherwise, including the build result (success or
+failure).
+
+I want to trigger a Jenkins or other CI system to take some action, e.g. run a
+test job, run another build, etc.
+
+To integrate with an external system that manages builds, e.g. building RPMs, I
+want to notify the external system that a build in OpenShift has started.
+
+I want to be able to perform some data extraction from the built image, to
+upload built binaries/assets to a server.
+
+
+### 3. Abort push
+
+As an application developer, I want to decide whether to push or not a image
+to a registry based on some customizable logic (e.g. test results).
+
+
+## Hooks
+
+1. Pre-build
+
+2. Pre-push
+
+3. Post-push
+
+4. Post-build
+
+
+## Alternatives
+
+Existing functionality that could be used to address the use cases above:
+
+- Kubernetes' container lifecycle hooks
+- `assemble` script (for S2I buils)
+- Changes in Dockerfile (for Docker build)
+- Deployment hooks
+- Events API and watchers
+- Custom build
+
+New functionality:
+
+- Build hooks
+
+
+## Discussion on alternatives
+
+Kubernetes' pre-stop hook is not a solution for any of the use cases, because
+it is only [executed prior to a SIGTERM when Kubernetes is about to terminate
+a container](https://github.com/GoogleCloudPlatform/kubernetes/blob/596a8a40d12498b5335140f50753980bfaea4f6b/docs/user-guide/production-pods.md#lifecycle-hooks-and-termination-notice).
+The builder container terminates itself, thus any pre-stop hook would
+be ignored.
```

Comment:
It is noted below how `assemble` does not apply to use cases "build pipeline" and "integration w/ external system".

Originally the structure of this section was to list each use case and discuss each potential solution for each build strategy, three dimensions.